col/src/main/java/vct/col/ast/util/NameScanner.scala

```diff
@@ -0,0 +1,210 @@
+    if (hasDecl(name)) {
+      Abort("Cannot put a free var when it has a decl")
+    }
+
+    Objects.requireNonNull(typ)
+
+    freeNames.get(name) match {
+      case Some(_) =>
+        // Disallow putting the same free var again, to prevent the writtenTo flag to be accidentally reset
+        Abort("Name %s is already marked as free", name);
+      case None =>
+        freeNames.put(name, Entry(name, typ, writtenTo = false));
+    }
+  }
+
+  private def setWrite(name: String): Unit =
+    freeNames.get(name) match {
+      case Some(value) => value.writtenTo = true
+      case None => Abort("Cannot set write for non-existent variable %s", name)
+    }
+
+  private def push(): Unit = frameStack.push(mutable.Set())
+
+  private def pop(): Unit = frameStack.pop()
+
+  /**
+    * Returns free names as a java map with insertion order retained
+    */
+  def freeNamesJava: util.Map[String, Type] = {
+    val map = new util.LinkedHashMap[String, Type]()
+    freeNames.foreach(mapEntry => map.put(mapEntry._1, mapEntry._2.typ))
+    map
+  }
+
+  /**
+    * All variable names that are read from or written to
+    * @return
+    */
+  def accesses: Set[String] = freeNames.keySet.toSet
+
+  def writes: Set[String] = freeNames.filter(p => p._2.writtenTo).keySet.toSet
+
+  def reads: Set[String] = freeNames.filter(p => !p._2.writtenTo).keySet.toSet
+
+  override def visit(e: NameExpression): Unit = e.getKind match {
+    case Reserved =>
+    case Label =>
+    case Field | Local | Argument =>
+      val name = e.getName
+      val t = e.getType
+
+      freeNames.get(name) match {
+        case Some(value) =>
+          // If the name is already a free variable, the types must match
+          if (t != null && !(value.typ == t)) {
+            Fail("type mismatch %s != %s", t, value.typ);
+          }
+        case None if !hasDecl(name) =>
+          // If it is new and there is no decl, declare a new free variable
+          put(name, t)
+        case _ =>
+          // Otherwise ignore it
+      }
+    case Unresolved =>
+      e.getName match {
+        case "tcount" =>
+        case "gsize" =>
+        case "tid" =>
+        case "gid" =>
+        case "lid" =>
+        case "threadIdx" =>
+        case "blockIdx" =>
+        case "blockDim" =>
+          put(e.getName, e.getType)
+      }
+    case _ =>
+      Abort("missing case %s %s in name scanner", e.getKind, e.getName)
+  }
+
+  override def visit(assignment: AssignmentStatement): Unit = {
+    super.visit(assignment)
+
+    assignment.location match {
+      case name: NameExpression if !hasDecl(name.name) => setWrite(name.name)
+      case _ =>
+    }
+  }
+
+  override def visit(d: DeclarationStatement): Unit = {
+    // Discovered a new declaration
+    frameStack.top.add(d)
+
+    super.visit(d)
+  }
+
+  override def visit(s: LoopStatement): Unit = {
+    push()
+
+    super.visit(s)
+
+    pop()
+  }
+
+  override def visit(e: BindingExpression): Unit = if (e.getDeclCount == 1) {
```

Comment:
Why this limitation?

Reply:
It was also in the original code, so that's why I left it in. Not sure if it has ever been a problem? Need to investigate a bit if it matters if we want to take it out

Reply:
Could not think/find any signs that this is still needed. I've changed it, will investigate if travis complains.